orchestra/machine_tasks.py

```diff
@@ -26,31 +32,46 @@ def execute(project_id, step_slug):
     # Machine tasks are only assigned to one worker/machine,
     # so they should only have one task assignment,
     # and should never be submitted for review.
-    count = task.assignments.count()
-    if count > 1:
-        raise MachineExecutionError('At most 1 assignment per machine task')
-    elif count == 1:
-        task_assignment = task.assignments.first()
-        if task_assignment.status == TaskAssignment.Status.SUBMITTED:
-            raise MachineExecutionError('Task assignment completed '
-                                        'but task is not!')
-    else:
-        task_assignment = (
-            TaskAssignment.objects
-            .create(task=task,
-                    status=TaskAssignment.Status.PROCESSING,
-                    in_progress_task_data={},
-                    snapshots={}))
+
+    with transaction.atomic():
+        # Uniqueness constraint on assignnment_counter and task prevents
+        # concurrent creation of more than one assignment
+        task_assignment, created = TaskAssignment.objects.get_or_create(
+            assignment_counter=0,
+            task=task,
+            defaults={
+                'status': TaskAssignment.Status.PROCESSING,
+                'in_progress_task_data': {},
+                'snapshots': {}})
+        if not created:
+            # Task assignment already exists
+            if task_assignment.status == TaskAssignment.Status.FAILED:
+                # Pick up failed task for reprocessing
+                task_assignment.status = TaskAssignment.Status.PROCESSING
+                task_assignment.save()
+            else:
+                # Task assignment already processing
+                raise MachineExecutionError(
+                    'Task already picked up by another machine')
 
     prerequisites = previously_completed_task_data(task)
 
     function_module = import_module(step.execution_function['module'])
     function = getattr(function_module, step.execution_function['name'])
-    task_data = function(project.project_data, prerequisites)
+    try:
+        task_data = function(project.project_data, prerequisites)
+    except:
+        task_assignment.status = TaskAssignment.Status.FAILED
+        logger.exception('Machine task has failed')
+        task_assignment.save()
+        return
     task_assignment.status = TaskAssignment.Status.SUBMITTED
     task_assignment.in_progress_task_data = task_data
     task_assignment.save()
-    task.status = Task.Status.COMPLETE
-    task.save()
-
-    create_subsequent_tasks(project)
+    if task.project.status == Project.Status.ABORTED:
```

Comment:
`# If a long-running task was aborted while running, we ensure the aborted state on the task.`